src/solidity-frontend/solidity_convert.cpp

```diff
@@ -2469,25 +2465,56 @@ bool solidity_convertert::get_binary_operator_expr(
     {
       // e.g. cmp(0x74,  0x7500)
       // ->   cmp(0x74, 0x0075)
-      exprt bwrhs, bwlhs;
-      bwrhs = exprt("bswap", rhs.type());
-      bwrhs.operands().push_back(rhs);
-      rhs = bwrhs;
+      assert(expr.contains("commonType"));
+      typet common_type;
+      if (get_type_description(expr["commonType"], common_type))
+        return true;
+
+      // convert string to bytes
+      // e.g.
+      //    data1 = "test"; data2 = 0x74657374; // "test"
+      //    assert(data1 == data2); // true
+      // Do type conversion before the bswap
+      // the arguement of bswap should only be int/uint type, not string
+      // e.g. data1 == "test", it should not be bswap("test")
+      // instead it should be bswap(0x74657374)
+      if (!is_bytes_type(lhs.type()))
+      {
+        if (get_expr(expr["leftExpression"], expr["commonType"], lhs))
```

Comment:
I wonder why parsing the expression again, but this time with a possibly different "common type" is necessary. It has already been translated (see comment above), no? If so, isn't the "convert_type_expr" happening below already taking care of casting to the common type?

Reply:
This is mainly for the string, (may also affect hex variable)
- So when not involving bytes, the string literal should be parsed as a char array by function `convert_string_literal()`
- When the string is used to initialise a byte type variable (`byte1 x = "a";`), the string literal should be parsed as a hex value by function `convert_hex_literal()`, where the corresponding hex value is already in the JSON node.
- The `literalType` is used to decide which function to call. Therefore we would call the get_expr again and overwrite the `lhs/rhs` if the type of the literal is byte.

The `convert_type_expr` does a different job, mainly handling the implicit typecast (e.g. `bytes16` => `bytes4`)

Reply:
I see, many thanks for this explanation! Maybe some of it could go into comments (if not already there).